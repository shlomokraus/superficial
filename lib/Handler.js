"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var Parser_1 = require("./Parser");
var file_extension_1 = __importDefault(require("file-extension"));
var Template_1 = require("./Template");
var lodash_1 = require("lodash");
var Github_1 = require("./Github");
var Persist_1 = require("./Persist");
var Constants_1 = require("./Constants");
var Handler = /** @class */ (function () {
    function Handler(context, pr) {
        var _this = this;
        this.parseFile = function (file) { return __awaiter(_this, void 0, void 0, function () {
            var valid, error, content, ex_1;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        valid = true;
                        _a.label = 1;
                    case 1:
                        _a.trys.push([1, 3, , 4]);
                        return [4 /*yield*/, this.getBaseAndHead(file)];
                    case 2:
                        content = _a.sent();
                        // If we don't have base or head then the file changes are real
                        valid =
                            content.base && content.head
                                ? this.compareFiles(content.head, content.base, file)
                                : true;
                        return [3 /*break*/, 4];
                    case 3:
                        ex_1 = _a.sent();
                        this.logger.error("Unable to parse file", ex_1);
                        error = true;
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, { file: file, valid: valid, error: error }];
                }
            });
        }); };
        this.revertFile = function (path) { return __awaiter(_this, void 0, void 0, function () {
            var base, source;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        base = this.pr.base.ref;
                        return [4 /*yield*/, this.githubHelper.getFile(path, base)];
                    case 1:
                        source = _a.sent();
                        return [2 /*return*/, { path: path, content: source.content }];
                }
            });
        }); };
        this.postRevertComment = function (files) { return __awaiter(_this, void 0, void 0, function () {
            var fileItemsTemplate, renderedFiles, commentTemplate, comment;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        fileItemsTemplate = Template_1.Template.get(Template_1.Templates.fileItem);
                        renderedFiles = files.map(function (filename) {
                            return lodash_1.template(fileItemsTemplate)({ filename: filename });
                        });
                        commentTemplate = Template_1.Template.get(Template_1.Templates.revert);
                        comment = lodash_1.template(commentTemplate)({
                            files: renderedFiles.join("\n")
                        });
                        comment = this.tagComment(comment, Constants_1.CommentTags.Update);
                        return [4 /*yield*/, this.githubHelper.createComment(comment)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); };
        this.postComment = function (items, errors) {
            if (errors === void 0) { errors = []; }
            return __awaiter(_this, void 0, void 0, function () {
                var files, comment, existing;
                return __generator(this, function (_a) {
                    switch (_a.label) {
                        case 0:
                            files = items.map(function (item) { return item.file; });
                            return [4 /*yield*/, this.compileComment(files, errors)];
                        case 1:
                            comment = _a.sent();
                            return [4 /*yield*/, this.getBotMainComment()];
                        case 2:
                            existing = _a.sent();
                            if (!existing) return [3 /*break*/, 4];
                            if (files.length === 0) {
                                comment = this.tagComment(Template_1.Template.get(Template_1.Templates.updated), Constants_1.CommentTags.Main);
                            }
                            return [4 /*yield*/, this.githubHelper.editComment(comment, String(existing.id))];
                        case 3:
                            _a.sent();
                            return [3 /*break*/, 6];
                        case 4:
                            if (!(files.length > 0)) return [3 /*break*/, 6];
                            return [4 /*yield*/, this.githubHelper.createComment(comment)];
                        case 5:
                            _a.sent();
                            _a.label = 6;
                        case 6: return [2 /*return*/];
                    }
                });
            });
        };
        this.getBotMainComment = function () { return __awaiter(_this, void 0, void 0, function () {
            var comments, filtered;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.githubHelper.getComments()];
                    case 1:
                        comments = _a.sent();
                        filtered = comments.find(function (comment) {
                            var isBot = comment.user.login === Constants_1.BOT_IDENTIFIER;
                            if (!isBot) {
                                return false;
                            }
                            var body = comment.body;
                            return body.indexOf(Constants_1.CommentTags.Main) >= 0;
                        });
                        return [2 /*return*/, filtered];
                }
            });
        }); };
        this.updateStatus = function (success) { return __awaiter(_this, void 0, void 0, function () {
            var state, message;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        state = success ? "success" : "failure";
                        this.logger.info("Updating status to " + state);
                        message = success
                            ? "ready to merge"
                            : "superficial changes not allowed";
                        return [4 /*yield*/, this.githubHelper.createStatus(state, message)];
                    case 1:
                        _a.sent();
                        return [2 /*return*/];
                }
            });
        }); };
        this.getBaseAndHead = function (path) { return __awaiter(_this, void 0, void 0, function () {
            var head, base, headRef, baseRef, ex_2;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        _a.trys.push([0, 3, , 4]);
                        headRef = this.pr.head.ref;
                        return [4 /*yield*/, this.githubHelper.getFileContent(path, headRef)];
                    case 1:
                        head = _a.sent();
                        baseRef = this.pr.base.ref;
                        return [4 /*yield*/, this.githubHelper.getFileContent(path, baseRef)];
                    case 2:
                        base = _a.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        ex_2 = _a.sent();
                        ex_2 = ex_2;
                        return [3 /*break*/, 4];
                    case 4: return [2 /*return*/, { head: head, base: base }];
                }
            });
        }); };
        this.context = context;
        this.logger = context.log;
        this.pr = pr;
        this.githubHelper = new Github_1.GithubHelper(this.context, pr);
        this.persist = new Persist_1.Persist(this.context, this.context.repo({ number: pr.number }));
    }
    /**
     * Main entry point
     */
    Handler.prototype.handle = function () {
        return __awaiter(this, void 0, void 0, function () {
            var event;
            return __generator(this, function (_a) {
                event = this.context.name ? this.context.name : this.context.event;
                this.logger.info("Handling action " + this.context.name);
                if (event === "issue_comment" &&
                    this.context.payload.action === "edited") {
                    return [2 /*return*/, this.handleCommentEdit()];
                }
                else {
                    return [2 /*return*/, this.handleCheckStatus()];
                }
                return [2 /*return*/];
            });
        });
    };
    Handler.prototype.handleCommentEdit = function () {
        return __awaiter(this, void 0, void 0, function () {
            var shouldRevert, files, revertPaths, revert;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.checkComment()];
                    case 1:
                        shouldRevert = _a.sent();
                        if (!this.context.payload.comment && !shouldRevert) {
                            this.logger.info("Comment is not a revert request");
                            return [2 /*return*/];
                        }
                        this.logger.debug("Reverting files");
                        return [4 /*yield*/, this.persist.get("files")];
                    case 2:
                        files = (_a.sent());
                        if (!files) {
                            this.logger.info("No files in metada: ", files);
                        }
                        this.logger.debug("Got files from metadata: ", files);
                        try {
                            files = JSON.parse(files);
                        }
                        catch (ex) {
                            this.logger.error("Error while parsing files to json", ex);
                        }
                        revertPaths = files.map(function (item) { return item.file; });
                        return [4 /*yield*/, Promise.all(revertPaths.map(function (path) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                                return [2 /*return*/, this.revertFile(path)];
                            }); }); }))];
                    case 3:
                        revert = _a.sent();
                        if (!(revert.length > 0)) return [3 /*break*/, 7];
                        this.logger.debug("Creating commit");
                        return [4 /*yield*/, this.githubHelper.createCommit(revert)];
                    case 4:
                        _a.sent();
                        this.logger.debug("Posting rever comment");
                        return [4 /*yield*/, this.postRevertComment(revert.map(function (file) { return file.path; }))];
                    case 5:
                        _a.sent();
                        return [4 /*yield*/, this.persist.set("files", undefined)];
                    case 6:
                        _a.sent();
                        _a.label = 7;
                    case 7: return [2 /*return*/];
                }
            });
        });
    };
    Handler.prototype.handleCheckStatus = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, problematic, errors, payload;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        this.logger.debug("Starting check run");
                        return [4 /*yield*/, this.check()];
                    case 1:
                        _a = _b.sent(), problematic = _a.problematic, errors = _a.errors;
                        this.logger.debug("Check completed with " +
                            problematic.length +
                            " files and " +
                            errors.length +
                            " errors");
                        this.logger.info("Updating pr status");
                        return [4 /*yield*/, this.updateStatus(problematic.length === 0)];
                    case 2:
                        _b.sent();
                        this.logger.info("Posting comment");
                        return [4 /*yield*/, this.postComment(problematic, errors)];
                    case 3:
                        _b.sent();
                        if (!(problematic.length > 0)) return [3 /*break*/, 5];
                        payload = JSON.stringify(problematic);
                        this.logger.info("Storing metadata", payload);
                        return [4 /*yield*/, this.persist.set("files", payload)];
                    case 4:
                        _b.sent();
                        _b.label = 5;
                    case 5: return [2 /*return*/];
                }
            });
        });
    };
    Handler.prototype.checkComment = function () {
        return __awaiter(this, void 0, void 0, function () {
            var context, changes, before, after, beforeCheck, afterCheck;
            return __generator(this, function (_a) {
                this.logger.info("Checking comment...");
                context = this.context;
                if (!this.context.payload.comment) {
                    this.logger.debug("No comment payload");
                    return [2 /*return*/, false];
                }
                if (this.context.payload.comment.user.login !== Constants_1.BOT_IDENTIFIER) {
                    this.logger.debug("Comment is not owned by bot");
                    return [2 /*return*/, false];
                }
                changes = context.payload.changes.body;
                if (!changes) {
                    this.logger.debug("No changes to comment body");
                    return [2 /*return*/, false];
                }
                before = changes.from;
                after = context.payload.comment.body;
                beforeCheck = before.indexOf("- [x] Remove selected files") < 0;
                afterCheck = after.indexOf("- [x] Remove selected files") >= 0;
                this.logger.debug("Before and after check:", beforeCheck, afterCheck);
                return [2 /*return*/, beforeCheck && afterCheck];
            });
        });
    };
    Handler.prototype.check = function () {
        return __awaiter(this, void 0, void 0, function () {
            var files, results, problematic, errors;
            var _this = this;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        this.logger.info("Getting files");
                        return [4 /*yield*/, this.githubHelper.getFiles()];
                    case 1:
                        files = _a.sent();
                        this.logger.info("Got " + files.length + " in pull request");
                        this.logger.debug(JSON.stringify(files));
                        files = this.filterFiles(files);
                        this.logger.info("Got " + files.length + " relevant files after filter");
                        return [4 /*yield*/, Promise.all(files.map(function (file) { return __awaiter(_this, void 0, void 0, function () { return __generator(this, function (_a) {
                                return [2 /*return*/, this.parseFile(file)];
                            }); }); }))];
                    case 2:
                        results = _a.sent();
                        this.logger.info("Finished parsing files, preparing results");
                        problematic = results.filter(function (result) { return !result.valid; });
                        errors = results.filter(function (result) { return result.error; });
                        return [2 /*return*/, { problematic: problematic, errors: errors }];
                }
            });
        });
    };
    Handler.prototype.compileComment = function (files, errors) {
        return __awaiter(this, void 0, void 0, function () {
            var fileItemsTemplate, renderedFiles, commentTemplate, comment, errorCount, errTemplate, errDoc;
            return __generator(this, function (_a) {
                fileItemsTemplate = Template_1.Template.get(Template_1.Templates.fileItem);
                renderedFiles = files.map(function (filename) {
                    return lodash_1.template(fileItemsTemplate)({ filename: filename });
                });
                commentTemplate = Template_1.Template.get(Template_1.Templates.comment);
                comment = lodash_1.template(commentTemplate)({
                    files: renderedFiles.join("\n")
                });
                errorCount = errors.length;
                if (errorCount > 0) {
                    errTemplate = Template_1.Template.get(Template_1.Templates.errors);
                    errDoc = lodash_1.template(errTemplate)({ errors: errorCount });
                    comment = comment + "\n" + errDoc;
                }
                return [2 /*return*/, this.tagComment(comment, Constants_1.CommentTags.Main)];
            });
        });
    };
    Handler.prototype.compareFiles = function (source, target, filename) {
        this.logger.info("Comparing file " + filename);
        var ext = file_extension_1.default(filename);
        var isScript = Constants_1.SCRIPT_EXTENSIONS.indexOf(ext) >= 0;
        var left = isScript
            ? Parser_1.Parser.parse(source, filename)
            : Parser_1.Parser.prepare(source, { filepath: filename });
        var right = isScript
            ? Parser_1.Parser.parse(target, filename)
            : Parser_1.Parser.prepare(target, { filepath: filename });
        var diff = Parser_1.Parser.diff(left, right, ["loc", "start", "end"]);
        this.logger.debug("Result is ", diff);
        return diff !== undefined;
    };
    Handler.prototype.filterFiles = function (files) {
        return files.filter(function (file) {
            var ext = file_extension_1.default(file);
            return Constants_1.VALID_EXTENSIONS.indexOf(ext) >= 0;
        });
    };
    /**
     * Tag comment with metadata
     */
    Handler.prototype.tagComment = function (body, commentTag) {
        if (body.indexOf(commentTag) < 0) {
            return body + "\n" + commentTag;
        }
        return body;
    };
    return Handler;
}());
exports.Handler = Handler;
