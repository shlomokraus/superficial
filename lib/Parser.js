"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var parser = require("@babel/parser");
var jsondiffpatch = require("jsondiffpatch");
var prettier = require("prettier");
var Parser = /** @class */ (function () {
    function Parser() {
    }
    Parser.parse = function (src, filename, prettier) {
        if (prettier === void 0) { prettier = true; }
        if (prettier) {
            src = this.prepare(src, { filepath: filename });
        }
        return parser.parse(src, {
            allowImportExportEverywhere: true,
            sourceType: "unambiguous",
            plugins: ["classProperties", "jsx", "typescript", "objectRestSpread", "decorators-legacy"]
        });
    };
    Parser.prepare = function (src, options) {
        return prettier.format(src, options);
    };
    Parser.diff = function (src, target, ignoreKeys) {
        if (ignoreKeys === void 0) { ignoreKeys = []; }
        var diff = jsondiffpatch.create({
            propertyFilter: function (name, context) {
                return ignoreKeys.indexOf(name) < 0;
            },
            objectHash: function (obj) {
                // this function is used only to when objects are not equal by ref
                if (obj.type === "ClassProperty") {
                    console.log(obj);
                    return "ClassProperty" + obj.key.name;
                }
                return null;
            },
        });
        return diff.diff(src, target);
    };
    return Parser;
}());
exports.Parser = Parser;
