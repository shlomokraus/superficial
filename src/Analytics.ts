import UA from "universal-analytics";

export class Analytics {
  private readonly ua;

  constructor(id?: string) {
    this.ua = new UA(id);
  }

  event(...args) {
    return this.ua.event.apply(this.ua, args).send();
  }

  exception(...args) {
    return this.ua.exception.apply(this.ua, args).send();
  }

  set(...args) {
    return this.ua.set.apply(this.ua, args);
  }

  timing(...args) {
    return this.ua.timing.apply(this.ua, args).send();
  }
}
