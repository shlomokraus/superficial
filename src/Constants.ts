export const BOT_IDENTIFIER = "superficial-bot[bot]";
export const VALID_EXTENSIONS = ["ts", "js", "tsx", "jsx", "json"];
export const SCRIPT_EXTENSIONS = ["ts", "js", "tsx", "jsx"];

export enum CommentTags {
  Main = "<!-- superficial-bot-main-comment -->",
  Update = "<!-- superficial-bot-update-comment -->"
}
