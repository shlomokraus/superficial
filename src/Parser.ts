const parser = require("@babel/parser");
const jsondiffpatch = require("jsondiffpatch");
const prettier = require("prettier");

export class Parser {
  static parse(src: string, filename: string, prettier = true) {
      if(prettier){
          src = this.prepare(src, { filepath: filename});
      }
    return parser.parse(src, {
      allowImportExportEverywhere: true,
      sourceType: "unambiguous",
      plugins: ["classProperties", "jsx", "typescript", "objectRestSpread", "decorators-legacy", "decorators"]
    });
  }

  static prepare(src: string, options) {
      return prettier.format(src, options);
  }

  static diff(src, target, ignoreKeys: string[] = []) {
    
    const diff = jsondiffpatch.create({
      propertyFilter: function(name, context) {
        return ignoreKeys.indexOf(name)<0;
      },
      objectHash: function(obj) {
        // this function is used only to when objects are not equal by ref
        if(obj.type==="ClassProperty"){
            console.log(obj);
            return "ClassProperty"+obj.key.name;
        }
        return null;
    },
    });
    return diff.diff(src, target);
  }

 
}
