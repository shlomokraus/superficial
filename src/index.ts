import { Handler } from "./Handler";
import { Application } from 'probot'

const events = [
    "pull_request.opened",
    "check_suite.requested",
    "check_suite.rerequested"
  ];

  export = async (app: Application) => {
  
    app.log("Starting superficial bot");
    app.log("Registering events", events);
    app.on(events, async (context) => {
        let prs;
        if(context.name==="pull_request" && context.payload.action==="opened"){
            prs = [context.payload.pull_request];
        } else {
            prs = context.payload.check_suite.pull_requests;
        }
        await Promise.all(prs.map(async pr => {
            const handler = new Handler(context, pr);
            return handler.handle();
        }))
       
    });

}

