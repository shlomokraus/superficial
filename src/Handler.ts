import { Context } from "probot";
import { Parser } from "./Parser";
import fileExtension from "file-extension";
import { Template, Templates } from "./Template";
import { template } from "lodash";
import { Github } from "./Github";
const extensions = ["ts", "js", "tsx", "jsx", "json"];
const botIdentifier = "superficialbot[bot]";
// const botId = 43469792;

export class Handler {
  private readonly context: Context;
  private readonly githubHelper;
  private readonly pr;

  constructor(context: Context, pr) {
    this.context = context;
    this.githubHelper = new Github(context, pr);
    this.pr = pr;
  }

  async handle() {
    const { problematic, errors } = await this.check();

    await this.updateStatus(problematic.length === 0);
    await this.postComment(problematic, errors);

    const revertPaths = problematic.map(item => item.file);
    const revert = await Promise.all(
      revertPaths.map(async path => this.revertFile(path))
    );
    if (revert.length > 0) {
        console.log(this.githubHelper.type);
   // await this.githubHelper.createCommit(revert);
   //   await this.postRevertComment(revert.map(file => file.path));

    }
  }

  async check() {
    const files = await this.getFiles();
    const results = await Promise.all(
      files.map(async file => this.parseFile(file))
    );
    const problematic = results.filter(result => !result.valid);
    const errors = results.filter(result => result.error);
    return { problematic, errors };
  }

  parseFile = async (file: string) => {
    let valid = true;
    let error;

    try {
      const content = await this.getBaseAndHead(file);
      // If we don't have base, that is a new file so it is valid
      valid =
        content.base && content.head
          ? this.compareFiles(content.head, content.base, file)
          : true;
    } catch (ex) {
      this.context.log.error("Unable to parse file", ex);
      error = true;
    }

    return { file, valid, error };
  };

  revertFile = async (path: string) => {
    const base = this.pr.base.ref;
    const source = await this.getFile(path, base);
    return { path, content: source.content };
  };

  postRevertComment = async (files: string[]) => {
    const fileItemsTemplate = await Template.get(Templates.fileItem);
    const renderedFiles = files.map(filename =>
      template(fileItemsTemplate)({ filename })
    );
    const commentTemplate = await Template.get(Templates.revert);
    const comment = template(commentTemplate)({
      files: renderedFiles.join("\n")
    });
    await this.context.github.issues.createComment(
      this.context.repo({ number: this.pr.number, body: comment })
    );
  };

  postComment = async (
    items: { file: string; valid: boolean }[],
    errors: { file: string; error: string }[] = []
  ) => {
    const context = this.context;
    const files = items.map(item => item.file);
    let comment = await this.compileComment(files, errors);

    const existing = await this.getExistingComment();
    if (existing) {
      await context.github.issues.editComment(
        context.repo({
          number: this.pr.number,
          body: `_(repository was updated since posting this message)_\n ${existing.body}`,
          comment_id: String(existing.id)
        })
      );
    } else {
      if (files.length > 0) {
        await context.github.issues.createComment(
          context.repo({ number: this.pr.number, body: comment })
        );
      }
    }
  };

  getExistingComment = async () => {
    const comments = await this.context.github.issues.getComments(
      this.context.repo({number: this.pr.number})
    );
    const filtered = comments.data.find(comment => {
      const isBot = comment.user.login === botIdentifier;
      if (!isBot) {
        return false;
      }

      const body = comment.body;
      return body.indexOf("be superficial!") >= 0;
    });
    return filtered;
  };

  async compileComment(files: string[], errors: any[]) {
    const fileItemsTemplate = await Template.get(Templates.fileItem);
    const renderedFiles = files.map(filename =>
      template(fileItemsTemplate)({ filename })
    );
    const commentTemplate = await Template.get(Templates.comment);
    let comment = template(commentTemplate)({
      files: renderedFiles.join("\n")
    });

    const errorCount = errors.length;
    if (errorCount > 0) {
      const errTemplate = await Template.get(Templates.errors);
      const errDoc = template(errTemplate)({ errors: errorCount });
      comment = comment + "\n" + errDoc;
    }
    return comment;
  }

  compareFiles(source: string, target: string, filename: string) {
    const left = Parser.parse(source, filename);
    const right = Parser.parse(target, filename);
    const diff = Parser.diff(left, right, ["loc", "start", "end"]);
    return diff !== undefined;
  }

  updateStatus = async (success: boolean) => {
    const { head } = this.pr;

    const state = success ? "success" : "failure";

    function getDescription() {
      if (success) return "ready to merge";
      return "superficial changes not allowed";
    }

    let status = {
      sha: head.sha,
      state,
      description: getDescription(),
      context: "Superficial"
    };
    await this.context.github.repos.createStatus(this.context.repo(
      status
    ) as any);
  };

  getFiles = async () => {
    const filesRaw = await this.context.github.pullRequests.getFiles(
      this.context.repo({ number: this.pr.number})
    );
    const files = filesRaw.data.map(file => file.filename);
    const filtered = this.filterFiles(files);
    return filtered;
  };

  getBaseAndHead = async (path: string) => {
    let head;
    let base;
    try {
      const headRef = this.pr.head.ref;
      head = await this.getFileContent(path, headRef);
      const baseRef = this.pr.base.ref;
      base = await this.getFileContent(path, baseRef);
    } catch (ex) {
      ex = ex;
    }

    return { head, base };
  };

  getFileContent = async (path: string, ref: string) => {
    const result = await this.getFile(path, ref);
    const content = Buffer.from(result.content, "base64").toString();
    return content;
  };

  getFile = async (path: string, ref: string) => {
    const repo = this.context.repo({ path, ref });
    const result = await this.context.github.repos.getContent(repo);
    return result.data;
  };

  filterFiles(files: string[]) {
    return files.filter(file => {
      const ext = fileExtension(file);
      return extensions.indexOf(ext) >= 0;
    });
  }
}
