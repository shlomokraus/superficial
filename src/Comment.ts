import { Template, Templates } from "./Template";
import { template } from "lodash";
import { CommentTags, BOT_IDENTIFIER } from "./Constants";
import { GithubHelper } from "./Github";

// const FILE_ITEM_REGEX = /- \[(.)\] (.*)<--file-->\|/g;

export class CommentHelper {
  private readonly githubHelper: GithubHelper;

  constructor(githubHelper) {
    this.githubHelper = githubHelper;
  }

  postComment = async (
    items: { file: string; valid: boolean }[],
    errors: { file: string; error: string }[] = []
  ) => {
    const files = items.map(item => item.file);
    let comment = await this.compileComment(files, errors);

    const existing = await this.getBotMainComment();
    if (existing) {
      if (files.length === 0) {
        comment = this.tagComment(
          Template.get(Templates.updated),
          CommentTags.Main
        );
      }
      await this.githubHelper.editComment(comment, String(existing.id));
    } else {
      if (files.length > 0) {
        await this.githubHelper.createComment(comment);
      }
    }
  };

  postRevertComment = async (files: string[]) => {
    const fileItemsTemplate = Template.get(Templates.fileItem);
    const renderedFiles = files.map(filename =>
      template(fileItemsTemplate)({ filename })
    );
    const commentTemplate = Template.get(Templates.revert);
    let comment = template(commentTemplate)({
      files: renderedFiles.join("\n")
    });

    comment = this.tagComment(comment, CommentTags.Update);
    await this.githubHelper.createComment(comment);
  };

  getBotMainComment = async () => {
    const comments = await this.githubHelper.getComments();
    const filtered = comments.find(comment => {
      const isBot = comment.user.login === BOT_IDENTIFIER;
      if (!isBot) {
        return false;
      }

      const body = comment.body;
      return body.indexOf(CommentTags.Main) >= 0;
    });
    return filtered;
  };

  async compileComment(files: string[], errors: any[]) {
    const fileItemsTemplate = Template.get(Templates.fileItem);
    const renderedFiles = files.map(filename =>
      template(fileItemsTemplate)({ filename })
    );
    const commentTemplate = Template.get(Templates.comment);
    let comment = template(commentTemplate)({
      files: renderedFiles.join("\n")
    });

    const errorCount = errors.length;
    if (errorCount > 0) {
      const errTemplate = Template.get(Templates.errors);
      const errDoc = template(errTemplate)({ errors: errorCount });
      comment = comment + "\n" + errDoc;
    }

    return this.tagComment(comment, CommentTags.Main);
  }

  tagComment(body, commentTag: CommentTags) {
    if (body.indexOf(commentTag) < 0) {
      return body + "\n" + commentTag;
    }
    return body;
  }
}
