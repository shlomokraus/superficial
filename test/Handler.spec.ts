const { Application } = require("probot");
import { Handler } from "../src/Handler";
import { Parser } from "../src/Parser";
const pr_sync_payload = require("./mocks/pull_request_open.json");

jest.mock("../src/Parser");
describe("Handler", () => {
  let context, issue, repo;

  describe("getBaseAndHead()", () => {
    beforeEach(() => {
      setupMocks();
    });

    it("Should get file from master and branch", async () => {
      const handler = new Handler(context as any);
      handler.getFileContent = jest
        .fn()
        .mockReturnValue(Promise.resolve("ABC"));
      const files = await handler.getBaseAndHead("test");
      const first = (handler.getFileContent as any).mock.calls[0];
      const second = (handler.getFileContent as any).mock.calls[1];
      expect(first).toEqual(["test", "test-style"]);
      expect(second).toEqual(["test", "master"]);
      expect(files.head).toEqual("ABC");
      expect(files.base).toEqual("ABC");
    });
  });

  describe("compareFiles()", () => {
    it("Should compare files and return true if diff is undefined", () => {
      const handler = new Handler(context as any);

      Parser.diff.mockImplementationOnce(() => undefined);
      const result = handler.compareFiles("source", "target", "test.ts");
      expect(Parser.parse).toBeCalledTimes(2);
      expect(Parser.parse.mock.calls[0]).toEqual(["source", "test.ts"]);
      expect(Parser.parse.mock.calls[1]).toEqual(["target", "test.ts"]);
      expect(result).toEqual(false);
    });

    it("Should false if diff is not undefined", () => {
      const handler = new Handler(context as any);

      Parser.diff.mockImplementationOnce(() => ({}));
      const result = handler.compareFiles("source", "target", "test.ts");
      expect(result).toEqual(true);
    });
  });

  describe("getFiles()", () => {
    beforeEach(() => {
      setupMocks();
    });

    it("Should extract files from pr", async () => {
      const handler = new Handler(context as any);
      const files = await handler.getFiles();
      expect(context.github.pullRequests.getFiles).toHaveBeenCalled();
      expect(files).toHaveLength(2);
      expect(files[0]).toEqual("test.ts");
      expect(files[1]).toEqual("test.js");
    });

    it("Should filter out non relevant extensions", async () => {
      const handler = new Handler(context as any);
      context.github.pullRequests.getFiles = jest.fn().mockReturnValue(
        Promise.resolve({
          data: [{ filename: "allow.ts" }, { filename: "block.exe" }]
        })
      )
      const files = await handler.getFiles();

      expect(files).toEqual(["allow.ts"]);
    });
  });
  
  describe("compileComment()", () => {
    beforeEach(() => {
      setupMocks();
    });

    it("Should filter out non relevant extensions", async () => {
      const handler = new Handler(context as any);
      const files = ["/test.ts", "/test.js"];
      const comment = await handler.compileComment(files);
      expect(comment).toMatchSnapshot();
    });
  });

  describe("updateStatus()", () => {
    beforeEach(() => {
      setupMocks();
    });

    it("Should post status update success", async () => {
      const handler = new Handler(context as any);
      await handler.updateStatus(true);
      expect(context.github.repos.createStatus).toHaveBeenCalled();
      expect(context.github.repos.createStatus.mock.calls[0][0].state).toEqual(
        "success"
      );
    });

    it("Should post status update failure", async () => {
      const handler = new Handler(context as any);
      await handler.updateStatus(false);
      expect(context.github.repos.createStatus).toHaveBeenCalled();
      expect(context.github.repos.createStatus.mock.calls[0][0].state).toEqual(
        "failure"
      );
    });
  });

  const setupMocks = () => {
    jest.clearAllMocks();
    issue = {
      owner: "tet",
      repo: "testing",
      number: 42
    };
    repo = {
      owner: "test",
      repo: "testing"
    };
    context = Object.assign(
      {
        log: () => {},
        repo: val => Object.assign(val, repo),
        github: {
          repos: {
            createStatus: jest.fn().mockReturnValue(Promise.resolve({}))
          },
          pullRequests: {
            getFiles: jest.fn().mockReturnValue(
              Promise.resolve({
                data: [{ filename: "test.ts" }, { filename: "test.js" }]
              })
            )
          }
        },
        issue: () => issue
      },
      pr_sync_payload
    );
  };
});
