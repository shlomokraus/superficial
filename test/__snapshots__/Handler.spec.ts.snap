// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Handler compileComment() Should filter out non relevant extensions 1`] = `
"### Don't be superficial!

This pull request contains files that have style changes without any functionality change. 
This makes the review and tracking of commit history harder. Please undo the changes to 
the following files: 

- /test.ts
- /test.js"
`;
